tmt/steps/__init__.py

```diff
@@ -392,12 +404,136 @@ def go(self):
         self.verbose('order', self.order, 'magenta', level=3)
 
 
-class Login(tmt.utils.Common):
-    """ Log into the guest """
+class ActionStep(tmt.utils.Common):
+    """ A special step which performs an action during a normal step. """
 
     # Dictionary containing list of requested phases for each enabled step
     _phases = None
 
+    @classmethod
+    def phases(cls, step):
+        """ Return list of phases enabled for given step """
+        # Build the phase list unless done before
+        if cls._phases is None:
+            cls._phases = cls._parse_phases(step)
+        # Return enabled phases, empty list if step not found
+        try:
+            return cls._phases[step.name]
+        except KeyError:
+            return []
+
+    @classmethod
+    def _parse_phases(cls, step):
+        """ Parse options and store phase order """
+        phases = dict()
+        options = cls._opt('step')
+
+        # Use the end of the last enabled step if no --step given
+        if not options:
+            last_enabled_step = list(step.plan.steps())[-1]
+            phases[last_enabled_step.name] = [PHASE_END]
+            return phases
+
+        # Process provided options
+        for option in options:
+            # Parse the step:phase format
+            matched = re.match(r'(\w+)(:(\w+))?', option)
+            if matched:
+                step_name, _, phase = matched.groups()
+            if not matched or step_name not in STEPS:
+                raise tmt.utils.GeneralError(f"Invalid step '{option}'.")
+            # Check phase format, convert into int, use end by default
+            try:
+                phase = int(phase)
+            except TypeError:
+                phase = PHASE_END
+            except ValueError:
+                # Convert 'start' and 'end' aliases
+                try:
+                    phase = dict(start=PHASE_START, end=PHASE_END)[phase]
+                except KeyError:
+                    raise tmt.utils.GeneralError(f"Invalid phase '{phase}'.")
+            # Store the phase for given step
+            try:
+                phases[step_name].append(phase)
+            except KeyError:
+                phases[step_name] = [phase]
+        return phases
+
+
+class Reboot(ActionStep):
+    """ Reboot guest """
+
+    # Which kind of reboot
+    _hard = False
+
+    # True if reboot enabled
+    _enabled = False
+
+    # Guest name provided via CLI
+    _guest_name = None
+
+    def __init__(self, step, order):
+        """ Initialize relations, store the reboot order """
+        super().__init__(parent=step, name='reboot')
+        self.order = order
+
+    @classmethod
+    def command(cls, method_class=None, usage=None):
+        """ Create the reboot command """
+        @click.command()
+        @click.pass_context
+        @click.option(
+            '-s', '--step', metavar='STEP[:PHASE]', multiple=True,
+            help='Log in during given phase of selected step(s).')
```

Comment:
Copy & paste typo. Interesting idea to support `step` selection. Originally I though the reboot feature would be only for ad hoc reboot requests, something like `tmt run --last reboot`. But this approach might come handy.

Reply:
I initially didn't expect it to turn out this way too but this way it offers a bit more flexibility, while also covering the most simple (and probably common) use case. Also it makes it convenient to share code between login and reboot.

Reply:
Should be solved.